File: LeetCode/98ValidateBinarySearchTree/ValidateBinarySearchTree.py
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        # DFS -> Time:O(n), Space:O(h)
        def valid(root, left, right):
            if not root:
                return True
            if left < root.val < right:
                return valid(root.left, left, root.val) and valid(root.right, root.val, right)
            return False
        return valid(root, float("-inf"), float("inf"))

File: LeetCode/98ValidateBinarySearchTree/ValidateBinarySearchTree.cpp
class Solution {
public:
    bool valid(TreeNode* root, long long left, long long right) {
        // DFS -> Time:O(n), Space:O(h)
        if (!root) {
            return true;
        }
        if (root->val > left && root->val < right) {
            return valid(root->left, left, root->val) && valid(root->right, root->val, right);
        }
        return false;
    }
    bool isValidBST(TreeNode* root) {
        return valid(root, LLONG_MIN, LLONG_MAX);
    }
};
